src/Features/CSharp/Portable/CodeFixes/AddParameterXmlDocumentation/AddParameterXmlDocumentationCodeFixProvider.cs

```diff
@@ -0,0 +1,316 @@
+                elementKind, out adjacentNameAttribute, out adjacentAttributeIsBefore);
+        }
+
+        private class FindAdjacentParameterDocumentationWalker : CSharpSyntaxWalker
+        {
+            private FindAdjacentParameterDocumentationWalker(int parameterIndex, Dictionary<string, int> parameterIndices, XmlNameAttributeElementKind elementKind)
+                : base(SyntaxWalkerDepth.StructuredTrivia)
+            {
+                _parameterIndex = parameterIndex;
+                _parameterIndices = parameterIndices;
+                _elementKind = elementKind;
+            }
+
+            private int _parameterIndex = -1;
+            private Dictionary<string, int> _parameterIndices = new Dictionary<string, int>();
+
+            private XmlNameAttributeSyntax _documentedParameterBefore = null;
+            private int _documentedParameterBeforeIndex = -1;
+            private XmlNameAttributeSyntax _documentedParameterAfter = null;
+            private int _documentedParameterAfterIndex = int.MaxValue;
+
+            private XmlNameAttributeElementKind _elementKind;
+
+            public static void FindAdjacentParameterDocumenteden(string parameterName, IEnumerable<string> allParameterNames,
+                SyntaxToken tokenWithDocumentationComment, XmlNameAttributeElementKind elementKind, out XmlNameAttributeSyntax adjacentNameAttribute, out bool adjacentAttributeIsBefore)
+            {
+                var parameterIndices = new Dictionary<string, int>();
+                int index = 0;
+                foreach (var param in allParameterNames)
+                {
+                    parameterIndices.Add(param, index);
+                    index++;
+                }
+                if (!parameterIndices.ContainsKey(parameterName))
+                {
+                    // something went very wrong
+                    Debug.Assert(false); // TODO message or remove assert
+                    adjacentNameAttribute = null;
+                    adjacentAttributeIsBefore = false;
+                    return;
+                }
+
+                var walker = new FindAdjacentParameterDocumentationWalker(parameterIndices[parameterName], parameterIndices, elementKind);
+                walker.VisitLeadingTrivia(tokenWithDocumentationComment);
+
+                if (walker._documentedParameterBefore != null)
+                {
+                    adjacentNameAttribute = walker._documentedParameterBefore;
+                    adjacentAttributeIsBefore = true;
+                }
+                else
+                {
+                    adjacentNameAttribute = walker._documentedParameterAfter;
+                    adjacentAttributeIsBefore = false;
+                }
+            }
+
+            public override void VisitXmlNameAttribute(XmlNameAttributeSyntax node)
+            {
+                if (node.GetElementKind() == _elementKind)
```

Comment:
I get that there's an enum, but is this anymore robust than just searching for "param"?